### DISCUSSION TITLE
IfcOpenShell python: Extracting 2D polygons from edges

### DISCUSSION THREAD

**Discussion by FransKnibbe** [IfcOpenShell/IfcOpenShell#4627] [Q&A] [answered]
Hello! I am trying to extract 2D polygons from IFC elements with ifcopenshell python. The 2D polygons should represent the ground footprint of the IFC element.

What is working so far: I can create a geometric shape from e.g. an IfcRoof element. The shape consists of vertices, edges and faces. From the vertices, I can create a convex hull that approximates the outer boundary.

In order to get an exact outer boundary, I think I need to create polygons based on the edges. I have been trying to manipulate the list of edges to create closed rings of line segments that can be used to define polygons, but that has not been entirely successful yet. 
As far as I can tell, the edges always consist of two indices: one for a 'from' vertex, and one for a 'to' vertex. But my understanding stops there. It would probably help to know how the edges are ordered by ifcopenshell.geom.create_shape(). Is there documentation on that?
And if someone has already found an algorithm for finding 2D polygons, I would love to hear about that!

Regards,
Frans

**Comment by Moult** [comment, marked as answer]
Have you looked at util.shape? It uses similar logic to getting real and projected footprint areas. You can then union the polygons with shapely.

Dion Moult

Sent from Proton Mail mobile

-------- Original Message --------

**Reply by FransKnibbe** [reply]
Do you mean look at the code for ifcopenshell.util.shape.get_footprint_area? I will do that. Hopefully I understand what is happening there...

**Reply by FransKnibbe** [reply]
Dion, thank you for that golden tip! Indeed util.shape.get_footprint_area has a good procedure for getting 2D footprints of IFC elements. I'm happy now!
